### PR TITLE
UniformGridLayout (FlowLayout): Fix issue with scrolling up creating gaps in layout

### LINKED ISSUE #1887: Setting `MaximumRowsOrColumns` property of `UniformGridLayout` causes holes and crashes in ItemsRepeater.
**Describe the bug**

Setting `MaximumRowsOrColumns` property of `UniformGridLayout` causes holes and crashes in `ItemsRepeater`. Holes appear in the upper right corner when using the latest stable version `2.3.191211002` of `Microsoft.UI.Xaml.Controls` on Nuget. Sliding errors with the `ScrollViewer` occur as well. The application crashed directly when scrolling the `ScrollViewer` using the latest prereleased version `2.4.0-prerelease.200113001`.

**Steps to reproduce the bug**

The test code is as follows
```xaml
<!--MainPage.xaml-->
<Page
    x:Class="TestApplication.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:TestApplication"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
    mc:Ignorable="d"
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Page.Resources>
        <!-- The Layout specifications used: -->
        <muxc:UniformGridLayout x:Name="UniformGridLayout"
                                MinRowSpacing="8" MinColumnSpacing="8"
                                MaximumRowsOrColumns="4"/>

        <!-- The DataTemplate used: CircularTemplate-->

        <DataTemplate x:Key="CircularTemplate" x:DataType="local:Bar">
            <Grid Background="Red"
                  Width="100"
                  Height="100">
                <TextBlock Text="{x:Bind str}" 
                           FontSize="20"/>
            </Grid>
        </DataTemplate>
    </Page.Resources>

    <ScrollViewer HorizontalScrollBarVisibility="Auto" 
        HorizontalScrollMode="Auto" 
        IsVerticalScrollChainingEnabled="False"
        MaxHeight="500">
        <muxc:ItemsRepeater
            ItemsSource="{x:Bind collection}"
            Layout="{StaticResource UniformGridLayout}"
            ItemTemplate="{StaticResource CircularTemplate}"
        />
    </ScrollViewer>
</Page>
```

```csharp
//MainPage.xaml.cs
using System;
using System.Collections.ObjectModel;
using Windows.UI.Xaml.Controls;

namespace TestApplication
{
    public sealed partial class MainPage : Page
    {
        public MainPage()
        {
            this.InitializeComponent();
            for (int i = 0; i < 40; ++i)
                collection.Add(new Bar(i.ToString()));
        }

        public ObservableCollection<Bar> collection
            = new ObservableCollection<Bar>();
    }

    public class Bar
    {
        public Bar(string str) { this.str = str; }

        public String str;
    }
}
```

**Screenshots**

a

**Version Info**

Windows 10 18363.592

NuGet package version: 

- `Microsoft.UI.Xaml 2.4.0-prerelease.200113001`
- `Microsoft.UI.Xaml 2.3.191211002`

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| November 2019 Update (18363)  |  Yes  |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :-------------------- | :------------------- |
| Desktop                 |   Yes |
| Mobile                   |    |
| Xbox                      |    |
| Surface Hub          |     |
| IoT                         |     |

### PR DESCRIPTION
**Pull request by chingucoding**
## Description

The issue was that the Generate(Backward) logic was faulty. When trying to generate the previous line, the next item was placed entirely to the right, however when having multiple items per row, this is not correct. That's why the UniformGridLayout ended up with single column content when scrolling up. The updated behavior is to take the last extent into account to place the last item of the previous row.

## Motivation and Context

Closes #1887 
## How Has This Been Tested?

Add new interaction test.

### CODE REVIEW

File: dev/Repeater/TestUI/Samples/UniformGridLayoutDemo.xaml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    <Page.Resources>
+        <!-- The Layout specifications used: -->
+        <controls:UniformGridLayout x:Name="UniformGridLayout"
+                                MinRowSpacing="8" MinColumnSpacing="8"
+                                MaximumRowsOrColumns="4"/>
+
+        <!-- The DataTemplate used: CircularTemplate-->
+
+        <DataTemplate x:Key="CircularTemplate" x:DataType="x:String">
```
Comment (ranjeshj) [inline dev/Repeater/TestUI/Samples/UniformGridLayoutDemo.xaml:16]: is this circular ?

File: dev/Repeater/TestUI/Samples/UniformGridLayoutDemo.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    public sealed partial class UniformGridLayoutDemo : Page
+    {
+        public ObservableCollection<string> collection
+            = new ObservableCollection<string>();
+
+        public UniformGridLayoutDemo()
+        {
+            this.InitializeComponent();
+            for (int i = 0; i < 40; ++i)
+                collection.Add(i.ToString());
```
Comment (ranjeshj) [inline dev/Repeater/TestUI/Samples/UniformGridLayoutDemo.cs:24]: could be simplified by just setting it in line 17.. using Enumerable.Range(0, 40)

File: dev/Repeater/TestUI/Samples/UniformGridLayoutDemo.xaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            </Grid>
+        </DataTemplate>
+    </Page.Resources>
+    <StackPanel Orientation="Horizontal">
+        <ScrollViewer HorizontalScrollBarVisibility="Auto" 
+            HorizontalScrollMode="Auto" 
+            IsVerticalScrollChainingEnabled="False"
+            AutomationProperties.Name="RepeaterScrollViewer"
+            MaxHeight="500">
+            <controls:ItemsRepeater x:Name="UniformGridRepeater"
```
Comment (ranjeshj) [inline dev/Repeater/TestUI/Samples/UniformGridLayoutDemo.xaml:31]: curious if this would be easier to see if you set vertical cache to 0

**Review by ranjeshj — APPROVED** [review]

**Comment by chingucoding** [comment]
@ranjeshj Would you like me to address your comments or rather merge as is?

**Comment by ranjeshj** [comment]
> @ranjeshj Would you like me to address your comments or rather merge as is?

Yes please. They were minor feedback, so if you can resolve them, that would be great and I don't expect those to cause new failures.

**Comment by chingucoding** [comment]
Added them now.

**Review by StephenLPeters — APPROVED** [review]
:shipit:

**Pull request merged**